tests/integrate_test/cita_discovery.sh

```diff
@@ -0,0 +1,403 @@
+                                         exit 1)
+
+    bin/cita bebop stop node/4 > /dev/null
+
+    # recover the config
+    for i in {0..3} ; do
+        sed '2d' -i node/$i/network.toml > /dev/null
+    done
+
+    echo "${timeout}s passed"
+}
+
+# T024: discovery node with max connected limit as server
+test_max_connected_limit_as_server() {
+    echo -n "T024 discovery node with max connected limit as server ... "
+
+    # stop all nodes
+    for i in {0..3} ; do
+        bin/cita bebop stop node/$i > /dev/null
+
+        # set max_connects = 2
+        sed '1 a\max_connects = 2' -i node/$i/network.toml > /dev/null
+    done
+
+    # start all nodes
+    for i in {0..3} ; do
+        bin/cita bebop start node/$i > /dev/null
+    done
+
+    # node[0..3]'s peer count is 2
+    for i in {0..3} ; do
+
+        # it is necessary to wait for a few seconds for each check
+        sleep 3
+        timeout=$(check_peer_count $i 2 90)||(echo "FAILED"
+                                              echo "error msg: ${timeout}"
+                                              exit 1)
+    done
+
+    # recover the config
+    for i in {0..3} ; do
+        sed '2d' -i node/$i/network.toml > /dev/null
+    done
+
+    echo "${timeout}s passed"
+}
+
+main() {
+    set -e
+
+    # FIXME: util needs this two global var, refine later
+    SOURCE_DIR=$(get_source_dir)
+    BINARY_DIR=${SOURCE_DIR}/target/install
+
+    echo -n "0) prepare  ...  "
+    # shellcheck source=/dev/null
+    . "${SOURCE_DIR}/tests/integrate_test/util.sh"
+    cd "${BINARY_DIR}"
+    set_hosts
+    echo "DONE"
+
+    echo -n "1) cleanup   ...  "
+    cleanup
+    echo "DONE"
+
+    echo -n "2) generate config  ...  "
+    generate_config
+    echo "DONE"
+
+    echo -n "3) pre start nodes[0..3]  ...  "
+    pre_start_nodes
+    echo "DONE"
+
+    # Pre-check peer count, it is OK to check node0 only
+    echo -n "4) pre-check peer count  ...  "
+    timeout=$(check_peer_count 0 3 60)||(echo "FAILED"
+                                         echo "error msg: ${timeout}"
+                                         exit 1)
+    echo "${timeout}s DONE"
+
+    test_suites=("ip" "domain_name")
+    index=0
+
+    for test_suit in "${test_suites[@]}" ; do
+        if [ "$test_suit" == "domain_name" ] ; then
+            # change node[4..7]'s peers to domain name
+            for i in {4..7} ; do
+                sed 's/127.0.0.1/node0/g' -i node/$i/network.toml
+            done
+        fi
+
+        test_node_entry ${index} 1
+        test_node_exit ${index} 2
+        test_parallel_entry ${index} 3
+
+        # this case should just after test_parallel_entry
+        test_parallel_exit ${index} 4
+        test_parallel_entry_exit ${index} 5
+        test_entry_before_known_node_start ${index} 6
+
+        index=$((index+1))
+    done
+
+    test_entry_and_known_node_restart
+    test_repeated_address
+#    test_max_connected_limit_as_client
```

Comment:
better to new one/two issue and then fix it on next pr.

Reply:
Okay!